controllers/notificationpolicy_controller.go

```diff
@@ -35,25 +37,40 @@ import (
 
 	"github.com/go-logr/logr"
 	"github.com/grafana/grafana-openapi-client-go/client/provisioning"
+	"github.com/grafana/grafana-operator/v5/api/v1beta1"
 	grafanav1beta1 "github.com/grafana/grafana-operator/v5/api/v1beta1"
 	client2 "github.com/grafana/grafana-operator/v5/controllers/client"
 )
 
+// LoopDetectedError is a custom error type for loop detection
+type LoopDetectedError struct {
+	Key string
+}
+
+func (e *LoopDetectedError) Error() string {
+	return fmt.Sprintf("loop detected, visited %s before", e.Key)
+}
+
 const (
-	conditionNotificationPolicySynchronized = "NotificationPolicySynchronized"
-	annotationAppliedNotificationPolicy     = "operator.grafana.com/applied-notificationpolicy"
+	conditionNotificationPolicySynchronized  = "NotificationPolicySynchronized"
+	conditionRoutesIgnoredDueToRouteSelector = "RoutesIgnoredDueToRouteSelector"
+	annotationAppliedNotificationPolicy      = "operator.grafana.com/applied-notificationpolicy"
 )
 
 // GrafanaNotificationPolicyReconciler reconciles a GrafanaNotificationPolicy object
 type GrafanaNotificationPolicyReconciler struct {
 	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
 }
 
 //+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationpolicies,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationpolicies/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationpolicies/finalizers,verbs=update
+//+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationpolicyroutes,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationpolicyroutes/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=grafana.integreatly.org,resources=grafananotificationpolicyroutes/finalizers,verbs=update
 
 func (r *GrafanaNotificationPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	r.Log = log.FromContext(ctx).WithName("GrafanaNotificationPolicyReconciler")
```

Comment:
#1836 was just merged which refactors how logging is propagated across the entire operator.
Would you try and adapt to the same changes?
```suggestion
	log = log.FromContext(ctx).WithName("GrafanaNotificationPolicyReconciler")
	ctc = log.IntoContext(ctx, log)
```

Reply:
I noticed and just did that 🙂